### PR TITLE
modified explorer data grid to follow discover look and feel

### PR DESCRIPTION
**Pull request by paulstn**
### Description
Adds data grid to log explorer that has the look and feel of the new Discover data grid
image

### Issues Resolved
Look and feel changes to Explorer Data Grid.
Known needed changes (p1-2 items): sidebar field order implementation, surrounding events implementation

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/components/event_analytics/explorer/sidebar/sidebar.tsx [hunk 1]
```diff
@@ -89,22 +93,34 @@ export const Sidebar = (props: ISidebarProps) => {
...
-      );
+      let nextFields;
+      if (
+        explorerFields.selectedFields.length === 0 &&
+        storedExplorerFields.selectedFields.length !== 0
+      ) {
+        nextFields = toggleFields(storedExplorerFields, field, SELECTED_FIELDS, AVAILABLE_FIELDS);
+      } else {
+        nextFields = toggleFields(explorerFields, field, SELECTED_FIELDS, AVAILABLE_FIELDS);
+      }
```
Comment (ps48) [inline public/components/event_analytics/explorer/sidebar/sidebar.tsx:121]: Minor: Can this be simplified? Suggesting something below (can be better)
```
    const selectedFields =
      explorerFields.selectedFields.length === 0 && storedExplorerFields.selectedFields.length > 0
        ? storedExplorerFields
        : explorerFields;

    const nextFields = toggleFields(selectedFields, field, SELECTED_FIELDS, AVAILABLE_FIELDS);

Comment (ps48) [reply public/components/event_analytics/explorer/sidebar/sidebar.tsx:121]: Also I see something similar in `handleAddField` do you want to create a function and reuse code?

File: public/components/event_analytics/explorer/events_views/doc_flyout.tsx [hunk 2]
```diff
@@ -124,13 +124,13 @@ export const DocFlyout = ({
...
       <div className="obsExplorer">
-        {populateDataGrid(
+        {/* {populateDataGrid(
           explorerFields,
           getHeaders(explorerFields.queriedFields, DEFAULT_COLUMNS.slice(1), true),
           <tr className="osdDocTable__row">{memorizedTds}</tr>,
           getHeaders(explorerFields.selectedFields, DEFAULT_COLUMNS.slice(1), true),
           <tr className="osdDocTable__row">{memorizedTds}</tr>
-        )}
+        )} */}
```
Comment (ps48) [inline public/components/event_analytics/explorer/events_views/doc_flyout.tsx:133]: Do you want to remove this commented piece?

File: public/components/event_analytics/explorer/events_views/data_grid.tsx [hunk 3]
```diff
@@ -3,148 +3,271 @@
...
+        }
+        return `${rows[trueIndex][columnId]}`;
+      }
+      return null;
+    },
+    [rows, pageFields, explorerFields]
+  );
+
+  // ** Pagination config
+  const [pagination, setPagination] = useState({ pageIndex: 0, pageSize: 25 });
```
Comment (ps48) [inline public/components/event_analytics/explorer/events_views/data_grid.tsx:219]: Is this default value same as discover?

Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:219]: Yes

File: public/components/event_analytics/explorer/events_views/doc_flyout.tsx (hunk 2 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/doc_flyout.tsx:133]: Will do

File: public/components/event_analytics/explorer/sidebar/sidebar.tsx (hunk 1 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/sidebar/sidebar.tsx:121]: Yep, will make the change now

File: public/components/event_analytics/explorer/__tests__/data_grid.test.tsx [hunk 4]
```diff
@@ -28,21 +37,44 @@ describe('Datagrid component', () => {
...
+      <Provider store={store}>
+        <DataGrid
+          http={http}
+          pplService={pplService}
+          rows={DATA_GRID_ROWS}
+          rowsAll={[]}
+          explorerFields={explorerFields}
+          timeStampField={'timestamp'}
+          rawQuery={
+            "source = opensearch_dashboards_sample_data_logs | where match(request,'filebeat')"
```
Comment (ps48) [inline public/components/event_analytics/explorer/__tests__/data_grid.test.tsx:61]: Minor: Might want to move this to test constants: https://github.com/opensearch-project/dashboards-observability/blob/main/test/event_analytics_constants.ts

Comment (paulstn) [reply public/components/event_analytics/explorer/__tests__/data_grid.test.tsx:61]: Moved

**Comment by derek-ho** [comment]
Can you add a screenshot of the new feel?

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 3 above)
Comment (derek-ho) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:219]: Seems like page size default for me is 100 - can we double check this/pull this value if it is set globall?

Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx]: You're right - changed the default page size to 100

File: public/components/event_analytics/explorer/events_views/data_grid.tsx [hunk 5]
```diff
@@ -3,148 +3,271 @@
...
+    storedSelectedColumns,
+  } = props;
+  const [rowRefs, setRowRefs] = useState<
+    Array<RefObject<{ closeAllFlyouts(openDocId: string): void }>>
+  >([]);
+  const { getEvents } = useFetchEvents({
+    pplService,
+    requestParams,
+  });
+  const sortingFields: MutableRefObject<GridSortingColumn[]> = useRef([]);
```
Comment (joshuali925) [inline public/components/event_analytics/explorer/events_views/data_grid.tsx:71]: why is this a ref? the component should rerender if sort changes

also type would be `MutableRefObject<EuiDataGridSorting['columns']>` if you use the upstream type

File: common/types/explorer.ts [hunk 6]
```diff
@@ -401,3 +401,8 @@ export type MOMENT_UNIT_OF_TIME =
   | 's'
   | 'milliseconds'
   | 'ms';
+
+export interface GridSortingColumn {
+  id: string;
+  direction: 'asc' | 'desc';
+}
```
Comment (joshuali925) [inline common/types/explorer.ts:408]: this is good but prefer to reuse existing type declaration `EuiDataGridSorting['columns']`, so we don't need to change type if upstream changes

File: public/components/event_analytics/explorer/events_views/data_grid.tsx [hunk 7]
```diff
@@ -3,148 +3,271 @@
...
+    redoQuery();
+  };
+
+  const setPage = (page: number[]) => {
+    pageFields.current = page;
+    redoQuery();
+  };
+
+  // creates the header for each column listing what that column is
+  const dataGridColumns = useMemo(() => {
```
Comment (joshuali925) [inline public/components/event_analytics/explorer/events_views/data_grid.tsx:117]: any perf issue needs useMemo? i don't think datagrid itself is memorized, not sure if useMemo/useCallback helps here

File: public/components/event_analytics/explorer/events_views/data_grid.tsx [hunk 8]
```diff
@@ -3,148 +3,271 @@
...
+      {
+        id: 'inspectCollapseColumn',
+        headerCellRender: () => null,
+        rowCellRender: ({ rowIndex }: { rowIndex: number }) => {
+          return (
+            <FlyoutButton
+              ref={null}
+              http={http}
+              key={null}
+              docId={'undefined'}
```
Comment (joshuali925) [inline public/components/event_analytics/explorer/events_views/data_grid.tsx:171]: why defining ref and key to be null? and why is docId `'undefined'`?

File: public/components/event_analytics/explorer/events_views/data_grid.tsx [hunk 9]
```diff
@@ -3,148 +3,271 @@
...
+        if (columnId === '_source') {
+          return (
+            <EuiDescriptionList type="inline" compressed>
+              {Object.keys(rows[trueIndex]).map((key) => (
+                <Fragment key={key}>
+                  <EuiDescriptionListTitle className="osdDescriptionListFieldTitle">
+                    {key}
+                  </EuiDescriptionListTitle>
+                  <EuiDescriptionListDescription
+                    dangerouslySetInnerHTML={{ __html: dompurify.sanitize(rows[trueIndex][key]) }}
```
Comment (joshuali925) [inline public/components/event_analytics/explorer/events_views/data_grid.tsx:201]: why is this needed rather than passing as children?

File: public/components/event_analytics/explorer/events_views/docViewRow.tsx [hunk 10]
```diff
@@ -36,7 +36,7 @@ interface IDocViewRowProps {
   onFlyoutOpen: (docId: string) => void;
 }
 
-export const DocViewRow = forwardRef((props: IDocViewRowProps, ref) => {
+export const FlyoutButton = forwardRef((props: FlyoutButtonProps, ref) => {
```
Comment (joshuali925) [inline public/components/event_analytics/explorer/events_views/docViewRow.tsx:39]: rename the file as well?

File: test/event_analytics_constants.ts [hunk 11]
```diff
@@ -571,3 +571,6 @@ export const HORIZONTAL_BAR_TEST_VISUALIZATIONS_DATA = {
 export const EXPLORER_START_TIME = 'Aug 28, 2023 @ 20:00:00.406';
 
 export const EXPLORER_END_TIME = 'Aug 28, 2023 @ 20:00:00.408';
+
+export const EXPLORER_DATA_GRID_QUERY =
+  "source = opensearch_dashboards_sample_data_logs | where match(request,'filebeat')";
```
Comment (joshuali925) [inline test/event_analytics_constants.ts:576]: unused?

**Review by joshuali925 — COMMENTED** [review]
minor comments

File: common/types/explorer.ts (hunk 6 above)
Comment (paulstn) [reply common/types/explorer.ts:408]: modified

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 7 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:117]: I was modeling this after Discover's data grid, and they have memoized table attributes. They also have their entire data grid memoized, would it be better if I did that as well? I assumed this was done to stop the rerendering of constant (barring user change) attributes such as columns but would this have no effect without the entire table memoization?

File: public/components/event_analytics/explorer/events_views/docViewRow.tsx (hunk 10 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/docViewRow.tsx:39]: I would, but more in depth flyout changes are slated for p1 so I was going to get to that (along with many other changes) later.

File: test/event_analytics_constants.ts (hunk 11 above)
Comment (paulstn) [reply test/event_analytics_constants.ts:576]: This is used in `public/components/event_analytics/explorer/__tests__/data_grid.test.tsx` as a specific raw query being passed into the data grid for testing.

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 9 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:201]: I assume the thinking was to sanitize the returned data before displaying to prevent XSS but I don't think it's needed here. I changed it back to passing the row value as a child

File: test/event_analytics_constants.ts (hunk 11 above)
Comment (joshuali925) [reply test/event_analytics_constants.ts:576]: move to test files if it's only used by tests?

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 9 above)
Comment (joshuali925) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:201]: sanitize is needed for `dangerouslySetInnerHTML`, i'm asking is it necessary to `dangerouslySetInnerHTML` in the first place? if not probably just pass as child

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 7 above)
Comment (joshuali925) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:117]: it can help if we run into performance issues, but right now useMemo doesn't really help since data grid is not memorized

Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:117]: Should I memoize data grid, remove useMemo in my code, or just leave it alone?

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 9 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:201]: done, i'm passing it in as a child.

File: test/event_analytics_constants.ts (hunk 11 above)
Comment (paulstn) [reply test/event_analytics_constants.ts:576]: Isn't this the test constant file? If not, where would it be?

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 7 above)
Comment (joshuali925) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:117]: i think you can remove or leave it if you don't see performance issues right now. we can add optimizations later when needed

File: test/event_analytics_constants.ts (hunk 11 above)
Comment (joshuali925) [reply test/event_analytics_constants.ts:576]: my bad, i misread the filename, you can ignore this comment..

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 7 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:117]: I'll leave it in, I believe we might need to include some performance optimizations in later

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 5 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:71]: Changed the type accordingly.
This needs to be a ref because the state won't persist across rerenders otherwise. I tested it out by removing the useRef() (and all associated) portions and an error occurs where the portion on the toolbar the says "Sort fields" doesn't keep track of which fields were sorted (and their directions).

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 8 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:171]: ref is a vestigial prop that will need to be removed later. 
key and docId are important, but for 'surrounding events', which as a whole isn't implemented yet but will be put in for p1. it won't have an effect on any other part of the flyout currently.

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 5 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx]: There is also an issue with having useState, it manifests itself as sort having a delay. Could be caused by getEvents, data cell rendering, or a few other things. Github Issue will be created after merge.

File: public/components/event_analytics/utils/__tests__/utils.test.tsx [hunk 12]
```diff
@@ -114,4 +116,27 @@ describe('Utils event analytics helper functions', () => {
...
+              direction: 'asc',
+            },
+          ],
+        },
+        {
+          current: [0, 100],
+        },
+        () => {}
+      )
+    ).toBe(undefined);
```
Comment (joshuali925) [inline public/components/event_analytics/utils/__tests__/utils.test.tsx:140]: i'm not sure how is expecting undefined validating the function?

you can probably at least do this
```suggestion
      const getEvents = jest.fn();
      redoQuery(
        'now/y',
        'now',
        "source = opensearch_dashboards_sample_data_logs | where match(request,'filebeat')",
        'timestamp',
        {
          current: [
            {
              id: 'timestamp',
              direction: 'asc',
            },
          ],
        },
        {
          current: [0, 100],
        },
        getEvents
      )
      expect(getEvents).toBeCalledWith(xxx)
```

File: public/components/event_analytics/explorer/events_views/data_grid.tsx [hunk 13]
```diff
@@ -3,148 +3,251 @@
...
+        rowCellRender: ({ rowIndex }: { rowIndex: number }) => {
+          return (
+            <FlyoutButton
+              ref={null}
+              http={http}
+              key={null}
+              docId={'undefined'}
+              doc={rows[rowIndex % pageFields.current[1]]}
+              selectedCols={explorerFields.queriedFields}
+              timeStampField={timeStampField}
```
Comment (mengweieric) [inline public/components/event_analytics/explorer/events_views/data_grid.tsx:141]: Should timeStampField be part of the useMemo dependency?

File: public/components/event_analytics/explorer/sidebar/sidebar.tsx [hunk 14]
```diff
@@ -87,24 +91,41 @@ export const Sidebar = (props: ISidebarProps) => {
     });
   };
 
+  const checkWithStoredFields = () => {
+    if (
+      explorerFields.selectedFields.length === 0 &&
+      storedExplorerFields.selectedFields.length !== 0
+    ) {
+      return storedExplorerFields;
+    } else {
```
Comment (mengweieric) [inline public/components/event_analytics/explorer/sidebar/sidebar.tsx:100]: else can be removed? only with return statement?

**Review by mengweieric — APPROVED** [review]
minor comments

File: public/components/event_analytics/utils/__tests__/utils.test.tsx (hunk 12 above)
Comment (paulstn) [reply public/components/event_analytics/utils/__tests__/utils.test.tsx:140]: I didn't think to have jest test for a dummy getEvents, this seems like it would test the function well. Thanks for suggesting this, I've implemented it.

File: public/components/event_analytics/explorer/events_views/data_grid.tsx (hunk 13 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/events_views/data_grid.tsx:141]: timeStampField is just a constant that is passed in as a prop from explorer.tsx. the intention for it being a prop is that if the user wants to define a time field instead of the default one, they can sort by that. but this is not implemented yet. I can still add it to the dependency array, though, for the future

File: public/components/event_analytics/explorer/sidebar/sidebar.tsx (hunk 14 above)
Comment (paulstn) [reply public/components/event_analytics/explorer/sidebar/sidebar.tsx:100]: removed else

**Review by joshuali925 — APPROVED** [review]
